### PR TITLE
PIG 11 - Light curves

### PR DESCRIPTION
**Pull request by dcfidalgo**
Hi,
I am really sorry for the big delay on this! Hopefully the draft is mature enough to initiate some discussion about improving the light curve computation in Gammapy.
I hope i will manage to add some psuedo code in form of a python script to this PR during the week.
I am very grateful for any kind of comments/suggestions/modifications, they are more than welcome!
David

### CODE REVIEW

File: docs/development/pigs/pig-xxx.rst [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+`gammapy.spectrum.SpectralModel` that is used to compute the expected counts in
+a time bin and to scale the integral flux with respect to the excess events
+found in that time bin.
+The parameters of the spectral model are generally obtained via a spectral fit
+to the whole data sample.
+A tutorial notebook of this approach can be found `here
+<http://docs.gammapy.org/dev/notebooks/light_curve.html>`_.
+
+Drawbacks of this approach are:
+ * the user has to estimate the spectral parameters beforehand, that is before
```
Comment (joleroi) [inline docs/development/pigs/pig-xxx.rst]: I'm not an expert in lightcurve analysis, but I always though the lightcurve point are basically flux points in time rather than energy space. So my understanding was that a global fit to the dataset always preceeds a lightcurve analysis and lateron only the model amplitude is adjusted in the time bins. In this propsal it sounds like a lightcurve analysis is always a full fit in a certain time interval. Can you specify somewhere which of the two options is the standard way?

File: docs/development/pigs/pig-xxx.rst [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+example by `gammapy.scripts.SpectrumAnalysisIACT`.
+The arguments for this pipeline should be provided in form of a *config dict*
+or possibly a txt file.
+One could think about providing a list of *config dict* or txt files to
+have more flexibility when performing the pipeline in each of the time bins.
+
+Instead of performing a spectral fit in each time bin, the user should also
+have the option to provide a fixed spectral shape, omit the fit and compute the
+integral flux the way it is done now, by scaling the expected counts with
+respect to the excess events, which should be much faster.
```
Comment (joleroi) [inline docs/development/pigs/pig-xxx.rst]: ok this answers my comment from above, maybe mention this earlier?

File: docs/development/pigs/pig-xxx.rst [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+time bin to disk, so that they can be easily read-in by the DataStore class
+with, for example, its `.from_dir()` method.
+This would avoid memory issues for large data sets and the user also gets the
+possibility to easily repeat a spectral analysis in one of the light curve
+bins.
+For this procedure `~gammapy.data.ObservationList` should get a method like 
+`.write_to_disk(outdir)`, or the `~gammapy.data.DataStore` object should get a
+method `.write_list_to_disk(obslist, outdir)` (invocation vs inspection).
+
+Depending on the modifications planned for the `gammapy.data.data_store`
```
Comment (joleroi) [inline docs/development/pigs/pig-xxx.rst]: As discussed at the coding sprint I would start with this

File: docs/development/pigs/pig-xxx.rst [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+class, e.g. we could think about extending the `gammapy.data.gti` class to
+provide some utilities for this.
+
+With this list/table the new `LightCurveEstimator` should then prepare/create a
+new `~gammapy.data.ObservatioList` for each time bin, in which the events have
+already been filtered properly and the
+`DataStoreObservation.observation_live_time_duration` were updated accordingly.
+These new `~gammapy.data.ObservationList` can then be passed on to the
+*standard* spectrum analysis with the `gammapy.background` and
+`gammapy.spectrum` modules.
```
Comment (registerrier) [inline docs/development/pigs/pig-xxx.rst]: I am not sure this logic is the best one. We don't really want to recreate a series of `Observation` for a specific time binning, no? This will become very heavy if you perform it on many different scales no?

Isn't it a simpler approach to add an event filter to the spectral extraction? 
After all this is not limited to time at all, one might want to perform a spectral analysis of some event types only. You don't want to recreate an `Observation` for this.

Then if all spectra have their own `TSTART`, `TSTOP` and `ONTIME` there is no difficulty extracting the light curve on any binning scheme build from the input one, by combining spectra during the fit process.

File: docs/development/pigs/pig-xxx.rst [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+have the option to provide a fixed spectral shape, omit the fit and compute the
+integral flux the way it is done now, by scaling the expected counts with
+respect to the excess events, which should be much faster.
+
+
+Preparing the ObservationList for each time bin
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The general idea is to write a new `~gammapy.data.ObservationList` for each
+time bin to disk, so that they can be easily read-in by the DataStore class
```
Comment (registerrier) [inline docs/development/pigs/pig-xxx.rst]: Same thing here.

File: docs/development/pigs/pig-xxx.rst [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+method `.write_list_to_disk(obslist, outdir)` (invocation vs inspection).
+
+Depending on the modifications planned for the `gammapy.data.data_store`
+module, one could also think about providing the user with the option to do
+everything in-memory, which could especially be useful for playing 
+around/testing.
+
+The preparation of the new `~gammapy.data.ObservationList` for each time bin
+would then basically consist of 3 steps:
+ * The `LightCurveEstimator` class figures out which observation belongs to
```
Comment (registerrier) [inline docs/development/pigs/pig-xxx.rst]: Observations might be cut into pieces as well no?

File: docs/development/pigs/pig-xxx.rst [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+
+Calling the spectral pipeline for each time bin
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+After preparing the observations and writing them to disk, a loop over the
+light curve bins should create a `~gammapy.data.DataStore` object -> 
+`~gammapy.data.ObservationList` and pass it on to the spectral pipeline (e.g.
+`gammapy.scripts.SpectrumAnalysisIACT`).
+Maybe the option to omit the spectral fit and scale a fixed spectrum to the
```
Comment (registerrier) [inline docs/development/pigs/pig-xxx.rst]: Again I think the logic is flawed here.

The first step is to produce reduced datasets and IRFs. So you simply want to perform spectral extraction. 

Spectral fitting should be done later so that any time rebidding can be performed on the fly.

This approach would force the user to rerun each time he want to change the bin size!

**Comment by registerrier** [comment]
I globally disagree with the approach where you have one single analysis step done by `LightCurveEstimator`.

There are 2 distinct things. 

1. data filtering and reduction where you:
- extract spectra and background in a given minimal time bin
- extract the corresponding reduced irfs namely _arf_ and _rmf_
2. data modeling and fitting where you:
- perform the spectral fit to your reduced datasets combining them to obtain the expected time binning (e.g. adaptative, minimal flux or whatever)

I also think that recreating Observations for the new time binning is too heavy an approach. It is simpler to add an event filtering to the spectral extraction and export the corresponding live time and start and stop time to the spectra (e.g. as metadata).

**Comment by cdeil** [comment]
After discussions at the Gammapy coding sprint today, @dcfidalgo and I started to brainstorm on how the time and other selections could work in Gammapy. See https://github.com/gammapy/gammapy-extra/blob/master/experiments/data.ipynb

I'm not sure if we should write a PIG.

There's a few concrete steps that are clear, and overall this is ~ 200 lines of code, and a working prototype is probably anyways needed to understand fully how it works and play with it. So I'd suggest we start sending PRs for this, adding to Gammapy without breaking what we have. (like we already did with the `ObservationCTA`).

**Comment by adonath** [comment]
@dcfidalgo If you want to keep this PIG in general and just update it later, it's fine to leave this PR open.
As PIGs are contained in individual files, we don't expect merge conflicts. Typically PRs with PIGs are merged when the PIG is either implemented or rejected, so as long as the PIG is updated it's fine to keep this PR open.

**Comment by dcfidalgo** [comment]
We just discussed a bit the future of this PIG in today's Gammapy call.
We decided to stop the development of this PIG, but leave the PR open to maybe serve as a starting point for future work on Light Curves.

**Comment by cdeil** [comment]
I have assigned this PIG a number and simplified the name: PIG 11 - Light curves
Authors are now Christoph Deil, Régis Terrier, Axel Donath, David Fidalgo, since David left.

I plan to update this PIG soon and then ask for review again.

For now, just very briefly: my suggestion would be to add two simple classes for 1D and 3D analysis that just script the existing analysis for 1D and 3D in time bins. Then we have the core functionality for Gammapy v1.0.

There will likely be background estimation issues if the time bins are small, and there a more complex analysis that creates background models from different time bins is needed - that's the motivation to keep the current LC maker that always does this on a per-run basis. Otherwise I would have suggested to just delete the existing code.

If we find someone that has an interest in lightcurves and works on this, we can improve the old and new LC maker further or try to integrate them to avoid duplication. But if that new person doesn't appear - we would still have the functionality for v1.0, just `gammapy.time` would remain one of the packages that's not cleaned up yet and would see further work after.

So basically: if anyone is interested to work on gammapy.time in the coming months - comment here or get in touch with @adonath @registerrier and me via Slack or email.

**Comment by bsipocz** [comment]
Would any of `astropy.timeseries` be useful here? Bear in mind, it's not feature complete at all, but now is a good time to suggest things for the API.

File: docs/development/pigs/pig-011.rst [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+events in the time bin considered, etc. In particular, for lightcurves with spectral
+variability studies, columns for other model parameters and associated errors would
+be necessary.
+The current container class already provides some methods to study variability, such as
+chi-square test, fractional variance estimation. But these methods are more general
+than the specific case of gammapy.
+
+This container is based on an ``astropy.table.Table`` object. It could therefore
+be improved by using the ``astropy_timeseries.BinnedTimeSeries`` object which
+itself inherits from ``Table`` and provides support for row selection with time,
```
Comment (bsipocz) [inline docs/development/pigs/pig-011.rst]: nitpicking:
```suggestion
itself inherits from ``QTable`` and provides support for row selection with time,
```

File: docs/development/pigs/pig-011.rst [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+Another complexity comes from the astrophysical background which can often pollute a source
+and needs to be properly removed to extract the intrinsic source flux at any given time.
+
+For these reasons, gamma-ray lightcurve are usually the results of a fit of model on
+the data performed on a number of time bins to extract the source flux in these time
+bins.
+
+This is more limited than e.g. time resolved spectral analysis. Although the latter share
+many similarities with the lightcurve extraction, it is a more complex task which
+we do not cover here.
```
Comment (adonath) [inline docs/development/pigs/pig-011.rst]: OK, to not cover it in this PIG, but maybe mention in an "outlook" section how it could be achieved? E.g. by allowing more free parameters for the spectral model, or e.g. run the `FluxPointsEstimator`  per time bin.

File: docs/development/pigs/pig-011.rst [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+
+    for dataset in datasets:
+        dataset.model = sky_model
+
+    lc_estimator = LightCurveEstimator(datasets)
+    lightcurve = lc.run()
+
+To help, the user deal with model parameters some helper function could be introduced
+to freeze all parameters of a model. Another possibility is to assume that all parameters
+are fixed except the `amplitude` in the ``LightCurveEstimator.run()`` method, or pass as
```
Comment (adonath) [inline docs/development/pigs/pig-011.rst]: For the `FluxPointEstimator` I have implemented a `reoptimize` argument. The basic logic is as following:
- With `reoptimize=False` only the amplitude of the spectral model is fitted.
- With  `reoptimize=True` all the free parameters of the dataset on init are fitted. This means the user specifies which parameters are frozen or not directly on the dataset objects (as for a "normal" likelihood fit).
I think this probably the simplest solution for now.

File: docs/development/pigs/pig-011.rst [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+Example usage:
+
+.. code ::
+
+    lightcurve.plot('amplitude')
+
+    # Get fractional variance
+    print(lightcurve.fvar('amplitude'))
+
+The ``LightCurve`` object should be able to rpovide some rebinning functionalities.
```
Comment (adonath) [inline docs/development/pigs/pig-011.rst]: I think this requires the likelihood profiles to be stored in the `LightCurve` object, which you currently don't mention yet.

File: docs/development/pigs/pig-011.rst [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+arguments the names of the parameters to leave free.
+
+Storing the results and further studies
+---------------------------------------
+
+The results are returned as a ``gammapy.time.LightCurve`` instance (the current
+container class for light curves) that, so far, essentially holds the integrated flux +
+errors and the time_min/time_max of each time bin. There are many other quantities
+which could be stored, such as the energy range of the flux, the number of excess
+events in the time bin considered, etc.
```
Comment (adonath) [inline docs/development/pigs/pig-011.rst]: Maybe be more specific here and give a list of quantities you propose to be stored in the `LightCurve`? In particular also mention the likelihood-profiles...

File: docs/development/pigs/pig-011.rst [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+
+    sky_model = SkyModel(spatial_model=spatial_model, spectral_model=spectral_model)
+    sky_model.parameters["index"].frozen = True
+    sky_model.parameters["lon_0"].frozen = True
+    sky_model.parameters["lat_0"].frozen = True
+
+    for dataset in datasets:
+        dataset.model = sky_model
+
+    lc_estimator = LightCurveEstimator(datasets)
```
Comment (adonath) [inline docs/development/pigs/pig-011.rst]: I think here you could add the additional arguments, we might want to have for the `LightCurveEstimator` such as `source="crab"`, maybe `reoptimize=True`, the norm values for the likelihood profiles etc.

One thing I was wondering is whether we should allow for `time_intervals` to be passed to the `LightCurveEstimator`.  The use case would probably monthly light-curves, where you do the data reduction per observation, pass all datasets to the `LightCurveEstimator`  and internally the datasets are grouped into monthly groups and the jointly fitted per group. I'm not sure it makes sense, but it's an idea, that came to my mind just now.

**Comment by cdeil** [comment]
I suggest we change https://docs.gammapy.org/0.14/notebooks/light_curve.html to use the high-level `Analysis` interface to get rid of the boilerplate code for data reduction.

I suggest to change [LightCurveEstimator](https://docs.gammapy.org/0.14/api/gammapy.time.LightCurveEstimator.html) to work like this:

```
time_bins = [....]
maker = LightCurveMaker(data, time_bins, config)
maker.run()
```
where `maker.run` does this:
```
results = []
for time_bin in self.time_bins:
    analysis = Analysis(... config, time_bins ...)
    # run data reduction, by default write to disk
    # run fit
    # store results
```
This has several advantages:
- Scales to many time bins, without running out of memory
- Simple to use, flexible via config.
- Easy to maintain, very little code
- We can probably support custom processing (e.g. for the fitting if there are other source components, or background) via hooks, i.e. advanced users sub-class LightCurveEstimator and override what some methods do (e.g. a global fit before extracting the UL for the spectral norm of interest)

My main gripe with https://docs.gammapy.org/0.14/notebooks/light_curve.html  is that as-written, if I process many time bins (say 100) in a 3D analysis, it will run out of memory.

@adonath @registerrier - I think you both prefer the double loop over time bins, first all data reduction, then all fitting. Is that still the case? If so, can you change the tutorial to show how that works without running our of memory for a 3D analysis so that I understand?

We should resolve the LC PIG and LC work for v1.0 in Nov.

**Comment by cdeil** [comment]
@registerrier - Thanks for finishing this up!

Please add the PIG to the toctree in [development/pigs/index.rst](https://github.com/gammapy/gammapy/blob/master/docs/development/pigs/index.rst). Usually when you then run `make docs-sphinx` you'll see formatting issues or broken links as Sphinx warnings. Please fix those also. 
Finally, please put a `Withdrawn: Oct 29, 2019` in the header of the PIG.
I see that this is also still missing [here](https://github.com/gammapy/gammapy/blob/master/docs/development/pigs/pig-007.rst) to put a date when it was withdrawn.

@registerrier - Could you please do this via a commit in master directly?

(and that's then the usual procedure always in the future for PIGs, to finish them up with such a commit adding it to the Sphinx docs and reviewing the HTML page once, either before or after merging)
